beacon-chain/types/crystallized_state_test.go

```diff
@@ -177,3 +177,108 @@ func TestProcessCrosslinks(t *testing.T) {
 		t.Errorf("Blockhash did not change for new cross link. Wanted a. Got: %s", newCrosslinks[0].Blockhash)
 	}
 }
+
+func TestIsDynastyTransition(t *testing.T) {
+	cState, err := NewGenesisCrystallizedState()
+	if err != nil {
+		t.Fatalf("Failed to initialize crystallized state: %v", err)
+	}
+	cState.data.DynastyStart = 1
+	if cState.IsDynastyTransition(0) {
+		t.Errorf("Is Dynasty transtion should be false, dynasty start greater than finalized slot")
+	}
+	cState.data.LastFinalizedSlot = 2
+	if cState.IsDynastyTransition(1) {
+		t.Errorf("Is Dynasty transtion should be false, MinDynastyLength has not reached")
+	}
+	shardCommitteeForSlots := []*pb.ShardAndCommitteeArray{{
+		ArrayShardAndCommittee: []*pb.ShardAndCommittee{
+			{ShardId: 0},
+			{ShardId: 1},
+			{ShardId: 2},
+		},
+	},
+	}
+	cState.data.ShardAndCommitteesForSlots = shardCommitteeForSlots
+
+	crosslinkRecords := []*pb.CrosslinkRecord{
+		{Slot: 1},
+		{Slot: 1},
+		{Slot: 1},
+	}
+	cState.data.CrosslinkRecords = crosslinkRecords
+
+	if cState.IsDynastyTransition(params.MinDynastyLength + 1) {
+		t.Errorf("Is Dynasty transtion should be false, crosslink records dynasty is higher than current slot")
+	}
+
+	crosslinkRecords = []*pb.CrosslinkRecord{
+		{Slot: 2},
+		{Slot: 2},
+		{Slot: 2},
+	}
+	cState.data.CrosslinkRecords = crosslinkRecords
+
+	if !cState.IsDynastyTransition(params.MinDynastyLength + 1) {
+		t.Errorf("Dynasty transition failed should have been true")
+	}
+}
+
+func TestNewDynastyRecalculationsInvalid(t *testing.T) {
+	cState, err := NewGenesisCrystallizedState()
+	if err != nil {
+		t.Fatalf("Failed to initialize crystallized state: %v", err)
+	}
+
+	// Negative test case, shuffle validators with more than MaxValidators.
+	var validators []*pb.ValidatorRecord
+	for i := 0; i < params.MaxValidators+1; i++ {
+		validators = append(validators, &pb.ValidatorRecord{StartDynasty: 0, EndDynasty: params.DefaultEndDynasty})
+	}
+	cState.data.Validators = validators
+	if _, err := cState.NewDynastyRecalculations([32]byte{'A'}); err == nil {
+		t.Errorf("Dynasty calculation should have failed with invalid validator count")
+	}
+}
+
+func TestNewDynastyRecalculations(t *testing.T) {
+	cState, err := NewGenesisCrystallizedState()
+	if err != nil {
+		t.Fatalf("Failed to initialize crystallized state: %v", err)
+	}
+
+	// Create shard committee for every slot.
+	var shardCommitteesForSlot []*pb.ShardAndCommitteeArray
+	for i := 0; i < params.CycleLength; i++ {
+		// Only 10 shards gets crosslinked by validators this dynasty.
+		var shardCommittees []*pb.ShardAndCommittee
+		for i := 0; i < 10; i++ {
+			shardCommittees = append(shardCommittees, &pb.ShardAndCommittee{ShardId: uint64(i)})
+		}
+		shardCommitteesForSlot = append(shardCommitteesForSlot, &pb.ShardAndCommitteeArray{ArrayShardAndCommittee: shardCommittees})
+	}
+
+	cState.data.ShardAndCommitteesForSlots = shardCommitteesForSlot
+	cState.data.CurrentDynasty = 1
+	cState.data.LastStateRecalc = 65
+
+	shardCount = 10
```

Comment:
Can you do a table driven test for this so we can test a few example shard counts?

https://github.com/golang/go/wiki/TableDrivenTests

Reply:
I'd like to see the default tested as well as 10 and a few other interesting numbers like 1, MaxInt64, 0, etc.